File: backend/app/schemas/mock.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel
from app.models.mock import RuleOperator, WhitelistType, MatchType


# MockRule schemas
class MockRuleBase(BaseModel):
    field: str
    operator: RuleOperator = RuleOperator.EQUALS
    value: str


class MockRuleCreate(MockRuleBase):
    pass


class MockRuleResponse(MockRuleBase):
    id: int
    suite_id: int

    class Config:
        from_attributes = True


# MockResponse schemas
class MockResponseBase(BaseModel):
    path: str
    method: str = "GET"
    response_json: Optional[str] = None
    ab_test_config: Optional[str] = None
    timeout_ms: int = 0
    empty_response: bool = False


class MockResponseCreate(MockResponseBase):
    pass


class MockResponseResponse(MockResponseBase):
    id: int
    suite_id: int

    class Config:
        from_attributes = True


# MockWhitelist schemas
class MockWhitelistBase(BaseModel):
    type: WhitelistType
    value: str


class MockWhitelistCreate(MockWhitelistBase):
    pass


class MockWhitelistResponse(MockWhitelistBase):
    id: int
    suite_id: int

    class Config:
        from_attributes = True


# MockSuite schemas
class MockSuiteBase(BaseModel):
    name: str
    description: Optional[str] = None
    is_enabled: bool = True
    enable_compare: bool = False


class MockSuiteCreate(MockSuiteBase):
    rules: List[MockRuleCreate] = []
    responses: List[MockResponseCreate] = []
    whitelists: List[MockWhitelistCreate] = []
    match_type: MatchType = MatchType.ANY


    class Config:
        from_attributes = True


class MockSuiteUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_enabled: Optional[bool] = None
    enable_compare: Optional[bool] = None
    rules: Optional[List[MockRuleCreate]] = None
    responses: Optional[List[MockResponseCreate]] = None
    whitelists: Optional[List[MockWhitelistCreate]] = None
    match_type: Optional[MatchType] = None


    class Config:
        from_attributes = True


class MockSuiteResponse(MockSuiteBase):
    id: int
    created_by: Optional[str] = None
    created_at: datetime
    updated_by: Optional[str] = None
    updated_at: datetime
    match_type: MatchType = MatchType.ANY
    rules: List[MockRuleResponse] = []
    responses: List[MockResponseResponse] = []
    whitelists: List[MockWhitelistResponse] = []

    class Config:
        from_attributes = True


class MockSuiteListResponse(BaseModel):
    total: int
    items: List[MockSuiteResponse]

File: frontend/src/api/mock.ts
import axios from 'axios'

const api = axios.create({
  baseURL: '/api/v1',
  timeout: 10000
})

export interface MockRule {
  id: number
  field: string
  operator: 'equals' | 'contains' | 'not_equals'
  value: string
}

export interface MockResponse {
  id: number
  path: string
  method: string
  response_json: string
  ab_test_config: string
  timeout_ms: number
  empty_response: boolean
}

export interface MockWhitelist {
  id: number
  type: 'clientId' | 'userId' | 'vid'
  value: string
}

export interface MockSuite {
  id: number
  name: string
  description: string
  is_enabled: boolean
  enable_compare: boolean
  created_by: string
  created_at: string
  updated_at: string
  match_type: 'any' | 'all'
  rules: MockRule[]
  responses: MockResponse[]
  whitelists: MockWhitelist[]
}

export interface MockSuiteListResponse {
  total: number
  items: MockSuite[]
}

export interface MockPreviewRequest {
  response_json: string
}

export interface MockPreviewResponse {
  valid: boolean
  formatted: string | null
  error: string | null
}

export const mockApi = {
  getSuites: (skip = 0, limit = 100) =>
    api.get<MockSuiteListResponse>('/mock/suites', { params: { skip, limit } }),
  getSuite: (id: number) =>
    api.get<MockSuite>(`/mock/suites/${id}`),
  createSuite: (data: Partial<MockSuite>) =>
    api.post<MockSuite>('/mock/suites', data),
  updateSuite: (id: number, data: Partial<MockSuite>) =>
    api.put<MockSuite>(`/mock/suites/${id}`, data),
  deleteSuite: (id: number) =>
    api.delete(`/mock/suites/${id}`),
  copySuite: (id: number, newName: string) =>
    api.post<MockSuite>(`/mock/suites/${id}/copy`, null, { params: { new_name: newName } }),
  previewResponse: (data: MockPreviewRequest) =>
    api.post<MockPreviewResponse>('/mock/preview', data),
}

export default mockApi
